expat/lib/expat.h

```diff
@@ -124,9 +124,64 @@ enum XML_Error {
   XML_ERROR_RESERVED_PREFIX_XMLNS,
   XML_ERROR_RESERVED_NAMESPACE_URI,
   /* Added in 2.2.1. */
-  XML_ERROR_INVALID_ARGUMENT
+  XML_ERROR_INVALID_ARGUMENT,
+  /* Added in 2.3.0 */
+  XML_ERROR_ENTITY_VIOLATION_SIZE,
+  XML_ERROR_ENTITY_VIOLATION_NESTED_SIZE,
+  XML_ERROR_ENTITY_VIOLATION_RATIO,
+  XML_ERROR_ENTITY_VIOLATION_DEPTH,
+  XML_ERROR_HASH_TABLE_VIOLATION
 };
 
+enum XML_Option {
+  /* Added in 2.3.0 */
+  XML_OPTION_HUGE_XML = 1, /* XML_Bool, no limits for huge XML (e.g. entity expansion) */
+  XML_OPTION_NESTING_LIMIT, /* int, limit entity nesting depth */
+  XML_OPTION_EXPANSION_RATIO, /* int, ratio between expansions and processed text */
+  XML_OPTION_MAX_EXPANSION_SIZE, /* XML_Size, limit max entity size for single and nested entities */
+  XML_OPTION_MAX_HASH_TABLE_ENTRIES /* XML_Size, max entries in DTD hash tables */
+};
+
+/* Entity expansion protection
+ *
+ * Mitigation against billion laugh and quadratic blowup attacks.
+ *
+ * XML_ENTITY_NESTING_LIMIT confines nesting of entities within entities.
+ * XML_ENTITY_EXPANSION_SIZE restricts the maximum length of entities,
+ *   both text of a single entity and resulting text of nested entities.
+ * XML_ENTITY_EXPANSION_RATIO constrains the ratio between position in XML
+ *  document and total amount of all expanded entities once more than
+ *  XML_ENTITY_EXPANSION_SIZE bytes have been expanded.
+ * XML_MAX_HASH_TABLE_ENTRIES limits total amount of entries across all
+ *   DTD hash tables.
+ *
+ * The limits are modelled after libxml2's limits
+ * https://github.com/GNOME/libxml2/blob/v2.9.8/parser.c#L99
+ * https://github.com/GNOME/libxml2/blob/v2.9.8/include/libxml/parserInternals.h#L33
+ */
+#ifndef XML_HUGE_XML_DEFAULT
+#define XML_HUGE_XML_DEFAULT 0
+#endif
+
+#ifndef XML_ENTITY_NESTING_LIMIT
+#define XML_ENTITY_NESTING_LIMIT 40
```

Comment:
Billion laughs demos use a nesting level of 10. Our default needs to be below 10 then.

Reply:
The settings limits the amount of nested expansions, not the maximum depths of an expansion stack. Perhaps we need a better name or explanation?

Reply:
Yes, we need a better name then. Some ideas mentioned further up.